contractcourt/channel_arbitrator.go

```diff
@@ -62,8 +62,13 @@ type WitnessBeacon interface {
 	// True is returned for the second argument if the preimage is found.
 	LookupPreimage(payhash []byte) ([]byte, bool)
 
-	// AddPreImage adds a newly discovered preimage to the global cache.
-	AddPreimage(pre []byte) error
+	// AddPreimages adds a batch of newly discovered preimages to the global
+	// cache, and also signals any subscribers of the newly discovered
+	// witness.
+	//
+	// NOTE: The backing slice of MUST NOT be modified, otherwise the
```

Comment:
It is a risk. You could use the new `lntypes.Preimage` type and go with copies.

Reply:
Thought about this, but we’d still be forced to perform a deep copy to ensure the notifications can’t be corrupted

Reply:
Nvm I see your point now, using `lntypes.Hash` makes a shallow copy sufficient. Good suggestion, also increases type safety. Will fix!

Reply:
yes, but then `lntypes.Preimage` for the preimages

Reply:
haha yes, that's what i meant to type :)

Reply:
👍

Reply:
I generally dig the move to use a `Preimage` type for type safety. Unfortunately, it turns out that the `channeldb.WitnessCache` is written in a general sense, meaning that it only operates on byte slices. So even though we get type safety going into the `preimageBeacon`, we then have to make a second copy of all of the preimages before passing them into the `WitnessCache`, so it ends up being pretty ugly.

One thing I've considered is making a wrapper/specialized interface, maybe `AddSha256Preimages` which uses the proper types, then these are serialized into the witness cache directly and hopefully avoiding the redundant copying. Lmk if I should investigate this further

Reply:
ended up going with the a helper that does only sha256 preimages, makes the code much cleaner